### PR TITLE
Use JLL binaries by default.

### LINKED ISSUE #328: Providing MPI through BinaryBuilder
I'm trying to get OpenMPI build through BB. For now, I have successful builds, but until we can try it out, we won't know what the issues are. Just posting here for folks who might be interested.

https://github.com/JuliaPackaging/Yggdrasil/pull/308

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #328 (as well as #373 and #376).

Summary:
 - Will use BinaryBuilder-provided MPICH / MicrosoftMPI by default
 - Provides override mechanism to use system MPI
 - Saves these preferences in `~/.julia/prefs/MPI.toml`

### CODE REVIEW

**Comment by ViralBShah** [comment]
Can we merge this?

**Comment by simonbyrne** [comment]
No: the JLL binaries work, but I'm not really satisfied with the mechanism for switching between the system MPI and JLL binaries.

Basically what this proposes is to use the system MPI (if found) and otherwise use the JLL, with overrides being set by various environment variables. There are a couple of problems with this approach:

1. We need to detect when the user changes their preferences, and invalidate the precompile cache. I still haven't figured out a robust way to do this (which is why the check is [currently disabled](https://github.com/JuliaParallel/MPI.jl/pull/367/files#diff-54ae667466f1993fc4117a62183ed265R75)).

2. Environment variables are probably not the best way to capture configuration options: they're inherently ephemeral, and users can forget to set them correctly.

3. Now that we're starting to get other JLLs that depend on MPI, using the system MPI by default will likely break those packages. e.g. Elemental.jl won't work if the user has a system MPI installed.

3 can be fixed by making using JLL the default, which I'm beginning to think might be the best option. 

I don't know how to deal with 1 and 2 though: ideally we would use the [Overrides.toml file](https://julialang.github.io/Pkg.jl/dev/artifacts/#Overriding-artifact-locations-1), but unfortunately it appears to be too limited to be of much use: you can only override the directory, and it assumes that all paths within that match the original artifact. Unfortunately this doesn't work for MPI (library and executable names can differ considerably). It appears we may need to develop some additional mechanism for overriding paths, but this would need to be used by all downstream JLLs as well, unless we can get the shim approach working.

**Comment by staticfloat** [comment]
> Unfortunately this doesn't work for MPI (library and executable names can differ considerably).

In this case, JLLs just aren't a good fit, IMO.  You're probably better off writing your own JLL-like package that uses artifacts directly when you need them, but has the additional flexibility to use other names.  JLLs are purposefully written in such a way as to expect something very consistent, an if names/SONAMEs are changing, I'm afraid they're the wrong tool for the job.

Ideally, we would have a layer that can absorb this complexity and provide a uniform interface to everything above it.  Is that even possible with so many different MPI implementations that we may need to deal with?

**Comment by simonbyrne** [comment]
We can do that, the main issue is that we would need to do the same for all downstream binaries as well, see https://github.com/JuliaParallel/Elemental.jl/pull/54

**Comment by staticfloat** [comment]
You mean that we would need different builds of downstream binaries because the downstream binaries will depend upon specific features within that flavor of MPI, or because the internal SONAMEs are different but the ABI is too different?

**Comment by ViralBShah** [comment]
> You mean that we would need different builds of downstream binaries because the downstream binaries will depend upon specific features within that flavor of MPI, or because the internal SONAMEs are different but the ABI is too different?

Yup - and these things will also want to link to their own BLAS / LAPACK. Some HPC sites require linking to their libraries.

**Comment by simonbyrne** [comment]
To clarify: I think it’s reasonable that if you’re using your own MPI (or BLAS), then you have to provide all your own downstream binaries as well. We just need a mechanism to do that.

**Comment by staticfloat** [comment]
> Yup - and these things will also want to link to their own BLAS / LAPACK. Some HPC sites require linking to their libraries.

I asked an "or" question, which one is it? ;)

**Comment by vchuravy** [comment]
>  because the internal SONAMEs are different but the ABI is too different?

That one. Difference in ABI and SONAMEs

There use to be things like morphmpi or wi4mpi (not an OSI license...) to bridge the ABI gaps

**Comment by simonbyrne** [comment]
Occasionally specific features as well (e.g. CUDA-aware interface)

**Comment by barche** [comment]
> I don't know how to deal with 1 and 2 though: ideally we would use the [Overrides.toml file](https://julialang.github.io/Pkg.jl/dev/artifacts/#Overriding-artifact-locations-1), but unfortunately it appears to be too limited to be of much use: you can only override the directory, and it assumes that all paths within that match the original artifact. Unfortunately this doesn't work for MPI (library and executable names can differ considerably). It appears we may need to develop some additional mechanism for overriding paths, but this would need to be used by all downstream JLLs as well, unless we can get the shim approach working.

I think that for some common use cases Overrides.toml could suffice, another option for really exotic uses is to simply roll out your own local JLL and put it in a local registry, [LocalRegistry.jl](https://github.com/GunnarFarneback/LocalRegistry.jl) makes that quite painless and its application to some common MPI implementations could be documented.

Another idea: provide JLLs for different common MPI implementations, but then we would need to have a way for the user to select one and the dependency-resolution would need to be aware that the same library can be provided by different packages. A bit like the ["provides" option](https://wiki.archlinux.org/index.php/PKGBUILD#provides) for Arch packages.

**Comment by ViralBShah** [comment]
Some MPI implementations are not easily available and are machine specific. And there's the cross product of each mpi implementation with each mpi library and blas.

**Comment by simonbyrne** [comment]
I've been thinking about this: what I think would work is something like the following:

1. A user config/preferences files (e.g. `~/.julia/prefs/MPI.toml`)
2. `Pkg.build("MPI")` would read the config file: if it exists it would generate a `deps/deps.jl` file that would define an appropriate `MPI_jll` module, and add a hook (using Requires.jl) that would throw an error if `MPICH_jll` is loaded.
3. The loading would read `deps/deps.jl` if it exists, otherwise `import MPICH_jll; const MPI_jll = MPICH_jll`.
4. `__init__()` would check that loading the config file gives the same answer.

All libraries using downstream binaries (e.g. Elemental.jl) would need to have the same mechanism for overloading, so it might be easier to create a lightweight package to do this.

**Comment by giordano** [comment]
> 2. `Pkg.build("MPI")` would read the config file: if it exists it would generate a `deps/deps.jl` file that would define an appropriate `MPI_jll` module, and add a hook (using Requires.jl) that would throw an error if `MPICH_jll` is loaded.
> 3. The loading would read `deps/deps.jl` if it exists, otherwise `import MPICH_jll; const MPI_jll = MPICH_jll`.

I think these steps can be slightly simplified, following the example of [`FFTW/deps/build.jl`](https://github.com/JuliaMath/FFTW.jl/blob/60e5169152ea4a899f19afa094950411c8644acb/deps/build.jl): `deps/build.jl` can be something like
```julia
if isfile("~/.julia/prefs/MPI.toml")
    # do something if the file exists
    # and generate the appropriate deps.jl file
else
    write("deps.jl", "using MPICH_jll, WindowsMPI_jll")
end
```
Ideally, in `MPI.jl` you shouldn't know what package had been loaded and it should only include `deps/deps.jl`

**Comment by simonbyrne** [comment]
Ah, I didn't realise FFTW did this, thanks!

**Comment by ViralBShah** [comment]
Note that many systems have a random MPI installed (from the linux distribution), and it would be preferable to use the JLLs and BB-compiled stuff.

**Comment by simonbyrne** [comment]
I can't get Travis to trigger for some reason...

**Comment by giordano** [comment]
https://travis-ci.org/github/JuliaParallel/MPI.jl/requests it says "Could not parse"

File: .travis.yml [hunk 1]
```diff
@@ -44,7 +52,7 @@ matrix:
   exclude:
   - os: osx
     env: MPI_IMPL=intelmpi
-
+bscribe" JULIA_DEBUG=MPI
```
Comment (giordano) [inline .travis.yml:55]: I guess this is the issue

**Comment by simonbyrne** [comment]
So to summarize:
1. By default, MPI.jl will use MicrosoftMPI_jll on Windows, MPICH_jll elsewhere
2. To use the system MPI, either:
  - set the environment variable `JULIA_MPI_BINARY=system`, or
  - add `binary="system"` to `~/.julia/prefs/MPI.toml` 

These preferences will be saved in the toml file between builds.

We will need a similar mechanism for all packages which load JLLs that depend on MPICH_jll (e.g. Elemental.jl).

Thoughts @lcw @vchuravy @barche @ViralBShah @giordano?

**Comment by barche** [comment]
I think this is a sensible way to deal with this, but some kind of "official" support in the JLL / package manager would be useful in the long term.

Minor detail: why not use the already existing `config` dir in `.julia` instead of `prefs`?

**Comment by simonbyrne** [comment]
> Minor detail: why not use the already existing `config` dir in `.julia` instead of `prefs`?

I picked it because that is what IJulia, FFTW and PyCall use, but I don't really mind.

**Comment by lcw** [comment]
Seems reasonable.  How is `mpirun`/`mpiexec` exposed to the user?

**Comment by giordano** [comment]
> We will need a similar mechanism for all packages which load JLLs that depend on MPICH_jll (e.g. Elemental.jl).

Why do they need a similar mechanism?  Won't they use whatever MPI is used by `MPI.jl`?

> Minor detail: why not use the already existing config dir in .julia instead of prefs?

I think using `prefs` was a decision of Steven Johnson for his packages.

**Comment by simonbyrne** [comment]
> Seems reasonable. How is `mpirun`/`mpiexec` exposed to the user?

There is a function for launching it from within Julia:
```
mpiexec(cmd -> run(`$cmd -n 4 foo`))
```
Unfortunately there is no great way to expose it outside of Julia (if you're using the system MPI, you can still use `mpiexec`/`mpirun` directly.)

**Comment by simonbyrne** [comment]
> Why do they need a similar mechanism? Won't they use whatever MPI is used by `MPI.jl`

Elemental.jl is switching over to use Elemental_jll (https://github.com/JuliaParallel/Elemental.jl/pull/54), which in turn [loads MPICH_jll](https://github.com/JuliaBinaryWrappers/Elemental_jll.jl/blob/b72537ed6bde7b1f52bc9cefc26153c6fef2aff3/src/wrappers/x86_64-linux-gnu-cxx11.jl#L5). It will need a mechanism to instead use a libelemental that has been built against the same MPI implementation that MPI.jl uses.

**Comment by giordano** [comment]
> Elemental.jl is switching over to use Elemental_jll (JuliaParallel/Elemental.jl#54), which in turn loads MPICH_jll. It will need a mechanism to instead use a libelemental that has been built against the same MPI implementation that MPI.jl uses.

Ah, I see.  Maybe it's worth documenting this?  I think there might be more JLL-ed libraries linked to MPI coming in future.

**Comment by PhilipVinc** [comment]
Hi, interested observer here. Sorry for the intrusion.

May I suggest that, if JLL Mpi is used, at MPI.jl build time you create a symlink for JLL's `mpirun/mpiexec` to something like `~/.Julia/bin`? You might even possibly rename them to `mpirun_jll`  so that that path could be safely added to the path....

**Comment by simonbyrne** [comment]
That’s a good idea, but I’m not sure how one would ensure it matches the correct artifact. @giordano any ideas?

**Comment by giordano** [comment]
Rather than a symlink to `mpiexec`/`mpirun` you probably want a wrapper script (Julia or bash?), as it'll need to know also where to find `libmpi`.  I may find some time this coming weekend to cook something for this.

**Comment by lcw** [comment]
Would the script have a `--project` flag to find the right version of MPI?  In other words, do we want to support having multiple versions of MPI.jl in `~/.julia` such that some point to the jll and some point to they system?

**Comment by giordano** [comment]
Can probably do that, yes.  My initial idea was to actually have something outside of `MPI.jl`, a small package like https://github.com/fredrikekre/jlpkg that would install the command line utility to run MPI.

**Comment by lcw** [comment]
I don't care if it is inside or out side MPI.jl but I am worried about keeping things in sync.

**Comment by simonbyrne** [comment]
Okay I have this working and have updated the docs. I'll leave the external `mpiexec` interface and dependent binary handling to other issues, since there is still stuff to be worked out there.

Please chime in if you have any more comments.

Thanks to @giordano and @staticfloat for tracking down the annoying Mac issues.

**Pull request merged**